lib/msal-node/src/client/ManagedIdentitySources/CloudShell.ts

```diff
@@ -3,17 +3,116 @@
+import { CryptoProvider } from "../../crypto/CryptoProvider";
+import {
+    HttpMethod,
+    METADATA_HEADER_NAME,
+    ManagedIdentityEnvironmentVariableNames,
+    ManagedIdentitySourceNames,
+    RESOURCE_BODY_OR_QUERY_PARAMETER_NAME,
+} from "../../utils/Constants";
+import {
+    ManagedIdentityErrorCodes,
+    createManagedIdentityError,
+} from "../../error/ManagedIdentityError";
 
+/**
+ * Original source of code: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/CloudShellManagedIdentitySource.cs
+ */
 export class CloudShell extends BaseManagedIdentitySource {
-    public createRequest(
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        _resource: string,
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        _managedIdentityId: ManagedIdentityId
-    ): ManagedIdentityRequestParameters {
-        throw new Error("Method not implemented.");
+    private msiEndpoint: string;
+
+    constructor(
+        logger: Logger,
+        nodeStorage: NodeStorage,
+        networkClient: INetworkModule,
+        cryptoProvider: CryptoProvider,
+        msiEndpoint: string,
+        systemAssigned: boolean
+    ) {
+        super(logger, nodeStorage, networkClient, cryptoProvider);
+
+        this.msiEndpoint = msiEndpoint;
+
+        if (!systemAssigned) {
+            throw createManagedIdentityError(
+                ManagedIdentityErrorCodes.unableToCreateCloudShell
+            );
+        }
+    }
+
+    public static tryCreate(
+        logger: Logger,
+        nodeStorage: NodeStorage,
+        networkClient: INetworkModule,
+        cryptoProvider: CryptoProvider,
+        systemAssigned: boolean
+    ): CloudShell | null {
+        const [areEnvironmentVariablesValidated, msiEndpoint]: [
+            boolean,
+            string | undefined
+        ] = validateEnvironmentVariables(
+            process.env[ManagedIdentityEnvironmentVariableNames.MSI_ENDPOINT],
+            logger
+        );
+
+        return areEnvironmentVariablesValidated
+            ? new CloudShell(
+                  logger,
+                  nodeStorage,
+                  networkClient,
+                  cryptoProvider,
+                  msiEndpoint as string,
+                  systemAssigned
+              )
+            : null;
+    }
+
+    public createRequest(resource: string): ManagedIdentityRequestParameters {
+        const request: ManagedIdentityRequestParameters =
+            new ManagedIdentityRequestParameters(
+                HttpMethod.POST,
+                this.msiEndpoint
+            );
+
+        request.headers[METADATA_HEADER_NAME] = "true";
+
+        request.bodyParameters[RESOURCE_BODY_OR_QUERY_PARAMETER_NAME] =
+            resource;
+
+        return request;
     }
 }
+
+const validateEnvironmentVariables = (
+    msiEndpoint: string | undefined,
+    logger: Logger
+): [boolean, string | undefined] => {
```

Comment:
is it needed to send a tuple here? It seems to be that a string carries this information anyway: 

- null (not undefined) string means "false" / MI is not available
- a value means "true"